Use Set as backing store for Utils.Bounded

coventry opened the pull request:
`Utils.Bounded` contains the comment

  `(* TODO one day replace list by an heap array *)`

`Set` uses balanced binary trees, so insertions are logarithmic in the size of theset, instead of linear as in the current list-based implementation.

In addition, this is a much simpler implementation.

In the process of writing this, I found a bug (#86) in the old implementation. This PR includes a test which the prior `Bounded` fails, in `test_utils.ml`. It's not plugged into the test Makefile framework, and I'm not sure how to do it. I'd be grateful for an explanation of how to do so.

Commit:
RPC: add HTTP methods and Content-Type headers.

Commit:
Client: add optional seed argument to Client_keys.gen_keys

Commit:
Utils: add [merge_list2] function

Commit:
RPC server: use ~on_exn

Commit:
Alphanet: dictator in the alpha protocol

Commit:
Genesis: unused error.

Commit:
Client: Add "network stat" command.

Commit:
Baker: fix operation ordering...

Commit:
Proto: classify and document some errors.

Commit:
typo

Commit:
Shell-level commit messages and timestamps.

Commit:
Validation: fetch new protocol on demand.

Commit:
Minor timing fix in the testsuite.

Commit:
Shell: fix `Utils.read_file`

Commit:
Shell: no duplicates in Prevalidation

Commit:
Shell: the genesis block should not flag the node as 'bootstraped'

Commit:
FunView: fix recursive removal.

Commit:
Update CI

Commit:
Client/genesis: option to set the timestamp of the activation block.

Commit:
Shell: avoid operation leaks

Commit:
RPC: add answer functions with HTTP code argument to RPC interface.

Commit:
Proto/Alpha: fix mining right timestamp

Commit:
Shell: fix (part of) the memory leak in validator

Commit:
Testsuite: fix error handling

Commit:
Makefile: add `--warn-error 8` (incomplete pattern-matching)

Commit:
CI: minor fix.

Commit:
Alpha: documentation of RPC wrappers.

Commit:
Client: avoid unrequired monitor

Commit:
Shell: fix the last known memory leak...

Commit:
Minor improvement in the docker image.

Commit:
Client: fix .merlin file.

Commit:
OPAM: more version contraint

Commit:
Alpha: classify and document contract errors.

Commit:
Alpha: classify and document qty errors.

Commit:
Alpha: classify and document mining errors.

Commit:
Client: do not fail when the protocol is unknown...

Commit:
Makefile: do not compile the (old) webclient

Commit:
Shell: add RPC to list network protocol versions.

Commit:
CI: fix public image

Commit:
P2p: implements peer swapping

Commit:
Typo

Commit:
Alpha: simplify mining/endorsement slots RPCs.

Commit:
Client: a bit of doc.

Commit:
CLient: fix `endorse for` command.

Commit:
Client: switch to JSON config file and remove config-file dependency.

Commit:
Client: lift some commands into functions.

Commit:
Client: add tags to aliases (enabled for contracts).

Commit:
Economic = economique
Economical = econome

Commit:
Alpha: classified public key revelation error.

Commit:
Utils: add [Lwt_utils.Idle_waiter].

Commit:
Shell: Merkle tree of operations

Commit:
Shell: repack the context every 10 commits.

Commit:
Switch to OCaml version 4.04.0.

Commit:
Alpha: fix typo in error registration.

Commit:
Alpha: take origination burn from source and not destination.

Commit:
Alpha: remove assets.

Commit:
Alpha: document and classify more errors.

Commit:
Client: add Client_proto_context.delegate_contract function

Commit:
RPCs: make the result of [/operations/X] compatible with Alpha's [/helpers/parse/operation].

fixes #160

Commit:
RPCs: make the result of [/blocks/X] compatible with Alpha's [/helpers/parse/block].

Commit:
Alphanet: add git to the docker image for [git repack] to work.

Commit:
Shell: fix unhandled exception in `io_scheduler`.

Commit:
Shell/Distributed_db: allow to `precheck` data.

Commit:
Alpha: remove unused storage key.

Commit:
[bugfix in data_encoding] When reading a float (from binary format), MBytes.get_double should be used instead of MBytes.get_float (Float _ encodes doubles in the encoding, not singles)

Commit:
[bugfix in data_encoding] When writing a float (to binary format), MBytes.set_double should be used instead of MBytes.set_float (Float _ encodes doubles in the encoding, not singles)

Commit:
add some asserts in data_encoding

Commit:
add more tests for "floats" in test_data_encoding ( +/- infinity, +/- 0., NaNs)

Commit:
Client/RPC: rename `wait` into `async`

Commit:
Shell: Split the operations list out of the (minimal) block header.

The minimal header now (classically) contains the root of a Merkle tree,
wrapping a list of lists of operations. Currently, the validator only
accept a single list of operations, but the 3+pass validator will
requires at least two lists.

Commit:
RPC: batch operation parsing/retrieval

Commit:
alpha: expose to_int32 for some types

Commit:
RPC: rebind 'complete' for blocks and operations.

Commit:
[Data_encoding.Binary] add a function 'to_bytes_list' to encode a given data as a list of 'small' blocks

Commit:
Client: do no pat parse the config file and `Sys.argv` twice.

Commit:
Shell: add unit tests for Merkle tree

Commit:
add an echo in ./test/test-basic.sh to say that the log of started node is dumped in ./test/LOG

Commit:
Merge branch 'improve-test-basic.sh' into 'master'

add an echo in ./test/test-basic.sh to say that the log of started node is dumped in ./test/LOG

See merge request !170

Commit:
Shell: Smaller Net_id.

Commit:
Merge branch 'merkle'

Commit:
Merge remote-tracking branch 'ocp/bugfix-in-data-encoding'

Commit:
Opam: use the `libev` backend of Lwt.

Commit:
Test: dump the content of the node logs.

Commit:
[data_encoding] use MBytes.{get/set}_char to read/write binary data instead of int8

Commit:
Merge branch 'use-MBytes-set-and-get-char' into 'master'

[data_encoding] use MBytes.{get/set}_char to read/write binary data instead of int8

See merge request !174

Commit:
Proto/Alpha: use record for mining slot.

Commit:
README.md update ocaml version to 4.04 (#9)

Commit:
Client: more `Error_monad` in `Client_{node,proto}_rpcs`

Commit:
Client: more `Error_monad` in `Client_aliases`

Commit:
Client: a bit of `Error_monad` in `Hash`

Commit:
Client: more `Error_monad` in `Cli_entries`.

Commit:
Test: simplify Makefile

Commit:
Client/Alpha: simplify `Client_proto_context`

Commit:
Proto/Alpha: some unit tests for 'transactions'

Commit:
Proto/Alpha: some unit tests for 'origination'

Commit:
Proto/Alpha: some unit tests for 'endorsement'

Commit:
Proto/Alpha: fix endorsement test

Commit:
Client: minor client output fixes and enhancements.

Commit:
RPC: fix CORS preflight.

Commit:
Shell/P2p: implements `raw_write_sync`.

For testing only: it allows to send 'gardled' messages to a peer.

Commit:
Shell/P2p: Split the unit tests in smaller atom.

And use more `Error_monad`...

Commit:
Shell: introduce `Error_monad._assert`.

Commit:
Shell/P2p: propagate all errors to `{raw_,}write_sync`.

Those functions are only used in the testsuite.

Commit:
Shell/P2p: fix connection close on read error

Commit:
Shell/P2p: propagate disconnection errors through the `pool`.

Commit:
Shell/P2p: add unit test with "garbled" messages.

Commit:
Proto: explicit fitness/timestamp in the signature

This remove the data fomr the context where they "duplicate" the
block header.

Commit:
Proto/Alpha: use `uint16` for mining priority.

Commit:
Shell/State: consistent use of `pred` vs. `predecessor`.

Commit:
Shell: remove dead code

Commit:
Shell: move `level` in th shell part of block.

Commit:
Shell/RPC: simplify block-watcher signature

Commit:
RPC: minor improvements in naming consistency

Commit:
Context: simplify the storage of 'test_network'.

This prepares the context to the inclusion the hash of the context in
the block header. By "looking" into the resulting context of a block,
we are now know able to determine whether:

- no testnet is currently associated to the branch;
- a testnet must be forked after the block;
- a previously forked testnet is running.

Commit:
Client: minor fix in the daemon.

Commit:
P2p: add test for invalid messages.

See merge request !175

Commit:
Split the signature on `Protocol.apply`.

This prepares the node for the multi-pass validator by allowing
more fine grained interaction between the shell and the economic
protocol.

See merge request !103

Commit:
Alpha: minor changes in unit tests

Commit:
P2p: remove compilation warning in unit test

Commit:
A few fixes to make things more posix compliant in build/test scripts

Commit:
Merge branch 'level_in_block_header' into 'master'

Move the `level` in the shell part of the block header

See merge request !177

Commit:
Shell: add the `protocol level` in the header of block

Commit:
Tests: add test for switching unanimously to 'demo' protocol

Commit:
data_encoding: read streams of Mbytes.t, tests, remove dead code (debug functions), expose a function that checks without reading the data

Commit:
Makefile: minor dependency fix in `test``

Commit:
Shell/Protocol: minor renaming

Commit:
Shell: lazy access to all the operations of a block

This prepares the node to the new on-disk storage.

Commit:
Shell/Proto: export more information to the RPC handler

Commit:
Merge branch 'stream-binary-decoding' into 'master'

[data_encoding] add facilities to decode a stream of Mbytes.t incrementally

See merge request !173

Commit:
Proto: add an RPC that return all the 'parsed' operations of a block

Commit:
test/utils: small fix: Data_encoding.Binary.Error has no arguments

Commit:
test_p2p_connection: put more deterministic garbled data

Commit:
Merge branch 'improve-tests' into 'master'

test_p2p_connection: put more deterministic garbled data

See merge request !180

Commit:
Proto/Alpha: replace Vote.`set`_current_proposal by `init`_current_proposal

Commit:
Proto/Alpha: fix proposal checking location in `Amendment.record_ballot`

Commit:
Proto/Alpha: add a RPC `/context/voting_period_kind`

Commit:
Docker: export "safe" RPCs from the container.

Commit:
Merge branch 'improved_rpc_context' into 'master'

Improved rpc context

See merge request !179

Commit:
Shell: do not keep tracks of 'invalid_successors'

Commit:
Test: inject operation and block atomically

Commit:
Docker: do not fail when there is nothing to backup...

Commit:
Proto: return the maximum TTL for operations

Commit:
Shell: implements `Chain.live_blocks`

Commit:
p2p: check values of some options in P2p.create

Commit:
P2p: use incremental encoding/decoding

Commit:
Docker: do not export the RPC port by default.

Commit:
RPCs: fix missing CORS headers.

Commit:
Shell: regroups basic data types in `Tezos_data`

Commit:
Shell: remove the on-disk index of operations

Let's get serious. The full index of operations is not sustainable in
the production code. We now only keep the index of operations not yet
in the chain (i.e. the mempool/prevalidation). Operations from the
chain are now only accesible through a block. For instance, see the
RPC:

   /blocks/<hash>/proto/operations

Commit:
Client: minor renaming

Commit:
Shell: Liveness of operations

Operations now include a block hash in their header. Such an operation
could only be included in a successor of this block.

Furthermore, when validating a block, the economic protocol now
returns---together with the context---an integer `max_operations_ttl`.
Then, when validating a successor, the shell will fail if it contains
an operation whose header's block hash is not one the
`max_operations_ttl` predecessors of the block.

As a bonus, the shell is now able to detect and forbid replayed
operations. Then, we might decide to remove some replay
detection-mechanism that we previously implemented in the economic
protocol.

Commit:
Tests: add voting period kind checks in `Test_vote`

Commit:
Proto/Alpha: minor renaming

Commit:
Proto/Alpha: more RPCs

Commit:
Shell/Proto: more complete "begin_construction".

The `begin_construction` function now accepts an optional argument
`proto_header`. This is to be used by a new RPC that ease forging the
shell header of a block (i.e. it will compute the fitness and, in a
near future, the hash of the resulting context).

Commit:
Shell: rework the `preapply` RPC

It now takes a `proto_header` in parameter, and it returns a full
`shell_header`. This prepares the inclusion of the context's hash in the
`shell_header`.

Commit:
Shell: minor renaming

Let's be consistent with Lwt (and ourselves).

Commit:
transfering -> transferring typo fix

Commit:
Fix compilation with lwt.3.0.0

Commit:
Client: add option for forcing free mining.

Commit:
Minor fix in contract deletion

Commit:
improve README: for installation of libsodium

Commit:
Merge branch 'improve-Readme-Install-Deps' into 'master'

improve README: for installation of libsodium

See merge request !186

Commit:
mention existence of the testnet

Commit:
Add: `tezos-client set delegate ...`

Commit:
do not pin cohttp to dev version

Commit:
Michelson: replace all numbers with only `int` and `nat`

Commit:
Shell: small bugfix preventing mempool cleanup.

Commit:
Alpha: add a faucet counter to prevent replay.

Commit:
Alpha: update bootstrap parameters.

Commit:
Shell: experimental leveldb backend

Commit:
Docker: quick fix for wrong upstream alpine docker.

Commit:
RPCs: fix node crash on SSL write error

Commit:
Alpha: consistent typechecking of contract storage initialization.

Commit:
Shell: export `Distributed_db.*.clear`

Commit:
Alpha: fix pre-execution typechecking.

Commit:
Revert "Alpha: add a faucet counter to prevent replay."

This reverts commit 512fde82e40d2a0956a7777cb3213db606e449e7.

Commit:
Client: derive the public key from the secret key when adding to wallet

That way we do not need to require the public key be added before hand
making the workflow nicer.

Commit:
Alpha: fix spelling of substraction

Commit:
Alpha: fix contract deletion.

Commit:
CI: reboot alphanet-next

Commit:
Doc: fixes incorrect usage example

Commit:
Alpha: adds an RPC to get the storage of a contract

Commit:
Michelson: adds missing instructions from language BNF

Commit:
Michelson: TRANSFER_TOKENS no longer takes a burn

Commit:
Michelson: allow CREATE_CONTRACT to set the spendable flag

Commit:
Update to mtime.1.0

Commit:
Update to irmin.1.3

Commit:
Michelson: updates spec to use `int` and `nat`

Commit:
Michelson: Adds DEFAULT_ACCOUNT instruction

Commit:
Michelson: adds DEFAULT_ACCOUNT test case

Commit:
Michelson: updates test suite to use `int` and `nat`

Commit:
Michelson: adds -amount flag to run program

Commit:
Michelson: show type_map for programs with type errors

Commit:
Michelson: removes NOP instruction from surface syntax

Commit:
Shell: document internal distributed_db error,

Commit:
Node: fix typos in help pages

Commit:
Node: fix typo in config variable name

Commit:
Michelson: pretty print typechecking errors when using run program

Commit:
Michelson: Quick and Dirty Emacs mode.

Commit:
Michelson: also display types inside lambdas.

Commit:
Michelson emacs-mode: auto resize type buffer.

Commit:
Michelson: removes amount from calling convention

Commit:
Alpha: fix error message for origination amount too low.

Commit:
Michelson: Add a test suite

Commit:
Alpha: adds a command to get the storage of a contract

Commit:
CI: run the Michelson test suite

Commit:
Michelson: add the .mli of the typechecker

Commit:
Michelson: better propagation and printing of errors to the client

Commit:
Michelson: fix typo in runtime error message

Commit:
Improves error messages and defaults for transfer -arg

Commit:
Alphanet: automatically copy files into container

Commit:
Makefile: fix `utop` rules

Commit:
Michelson: adds SIZE for sets and maps

Commit:
Michelson-mode: async, cache, and refactorings

Commit:
Update to ocaml.4.04.2

Commit:
Fix .dockerignore

Commit:
Update CI

Commit:
Prevalidation: minor bugfix.

Commit:
Michelson: adds tests for DEFAULT_ACCOUNT

Commit:
Update protocol version

Commit:
CI: fix the 'deploy' rules of alphanet/zeronet

Commit:
Testing: Tests multiple nodes

Commit:
CI: use our own docker image of `opam`

Commit:
Add `scripts/README.alphanet`.

Commit:
Adds sandbox launching script

Commit:
doc typo set::MEM (again)

This pull request is identical to recently closed #19, but applied to the current head. I think #19 may have been closed erroneously. If MEM takes as an input a set of elements of type 'elt, then if it also takes another input which is an element to test for membership, clearly that other input must be of type 'elt.

If #19 was not closed erroneously, then please explain, and i will stop reopening this issue.

Commit:
Add `CHANGES.alphanet`.

Commit:
Node: fix misplaced logging message

Commit:
Emacs: vertical stack printing and suggestions

Commit:
scripts: Set 'echo -e' to make text colors work in alphanet.sh

Before this change, output from `alphanet.sh` looked like the following:

```
New contract alias 'my_account' saved.
Contract memorized as my_account.
\033[32mThe node is now running.\033[0m
\033[32mThe baker is now running.\033[0m
\033[32mThe endorser is now running.\033[0m
\033[33mWarning: the container contains a new version of 'alphanet.sh'.\033[0m
\033[33mYou might run './scripts/alphanet.sh update_script' to synchronize.\033[0m
```

I.e. control characters showed up, instead of modifying the color of the text
output. Repro'd on Container Linux (CoreOS) 1492.3.0.

After this change, `The node is now running` shows up in green &c.

Commit:
Fix bug in launch node script

Commit:
Michelson: Fixes reversed list bug

Commit:
CI: minor tweaks

Commit:
Michelson: Enforces ordering on data

Commit:
Reset the chain (on-disk storage updated)

Commit:
CI: minor bugfix

Commit:
Michelson: Add LAMBDA to the language spec

Commit:
Utils: fix `list_rev` complexity.

Using fold_left here iterates over the entire list, whereas you only
need to iterate over the first `n` elements.

Commit:
Update Readme.md

Commit:
Michelson: remove a spurious rule

This should have been removed when floats were removed from the
language.

Commit:
Makefile: fix `node.top`

coventry commented:
Added a test which the old `Bounded` fails on, updated the header comment.

Commit:
Utils: various doc improvements

Commit:
Michelson: small documentation fix-up.

Commit:
Alphanet: fixes `container:` relative paths bug

Commit:
Adds test for default account creation/deletion

Commit:
Makefile: fix compat issue with macOS.

Commit:
P2p: comment out dead code

Commit:
Utils: fix `Utils.take_n` in presence of duplicates

OCamlPro-Henry commented on src/minutils/utils.ml:
```diff
@@ -206,35 +206,22 @@ let take_n_unsorted n l =
 
 module Bounded(E: Set.OrderedType) = struct
 
-  (* TODO one day replace list by an heap array *)
```
That comment is refering to : https://en.wikipedia.org/wiki/Binary_heap

ghost commented:
About the test suite, see https://github.com/tezos/tezos/commit/c091add3352cf6563d2a46bdc4196dd402cf226d for an example.

OCamlPro-Henry commented on src/minutils/utils.ml:
```diff
@@ -206,35 +206,22 @@ let take_n_unsorted n l =
-    end else if E.compare (List.hd t.data) x < 0 then
-      t.data <- replace x t.data
+    t.data <- SE.add x t.data;
+    if (t.bound < SE.cardinal t.data) then
```
The current implementation of `cardinal` traverses the whole set.

coventry commented:
Thanks for the feedback.  It's surprising that `cardinal` works like that. I can add a `count` field to `Bounded` to work around that. Also, a `min` field to track the smallest element, to work around the logarithmic lookup time in `Set`.

coventry committed:
Use `Set` as backing store for `Utils.Bounded`

`Set` uses balanced binary trees, so insertions are logarithmic in the size of
the set, instead of linear as in the current list-based implementation. The
`size` field keeps track of the backing set so that `Set.cardinal`, which has
run-time complexity linear in the size of the set, can be avoided

In addition, this is a much simpler implementation.

Includes a test of Bounded based on general invariants

coventry commented:
I added those features to the algorithm. It should have the same asymptotic runtime as the heap array implementation, now. If it's over-engineering, I understand. It was good practice.

hnrgrgr commented:
Closing all MR in github, please use https://gitlab.com/tezos/tezos instead.
